lib/phoenix_live_view/colocated_hook.ex

```diff
@@ -0,0 +1,232 @@
+      if !String.contains?(manifest, js_filename) do
+        IO.binwrite(
+          file,
+          ~s|\nimport hook_#{js_filename} from "#{js_full_path}"; hooks["#{js_filename}"] = hook_#{js_filename};|
+        )
+      end
+    end)
+
+    %{hashed_name: js_filename}
+  end
+
+  @doc false
+  def prune(hook, hashed_name) do
+    cond do
+      hook.opts[:bundle_mode] == :runtime ->
+        throw(:noop)
+
+      hook.opts[:bundle_mode] == :current_otp_app and not ColocatedHook.should_bundle?(hook) ->
+        throw(:noop)
+
+      true ->
+        :ok
+    end
+
+    manifest_path = manifest_path(hook)
+    dir = Path.dirname(manifest_path)
+
+    case File.ls(dir) do
+      {:ok, hooks} ->
+        for hook_basename <- hooks do
+          if String.starts_with?(hook_basename, hashed_name) do
+            File.rm!(Path.join(dir, hook_basename))
+
+            if File.exists?(manifest_path) do
+              new_file =
+                manifest_path
+                |> File.stream!()
+                |> Enum.filter(fn line -> !String.contains?(line, hashed_name) end)
+                |> Enum.join("")
+                |> String.trim()
+
+              File.write!(manifest_path, new_file)
+            end
+          else
+            :noop
+          end
+        end
+
+      _ ->
+        :noop
+    end
+  catch
+    :noop -> :noop
+  end
+end
+
+defimpl Phoenix.LiveView.TagExtractor, for: Phoenix.LiveView.ColocatedHook do
+  alias Phoenix.LiveView.ColocatedHook
+
+  def extract(
+        %Phoenix.LiveView.ColocatedHook{opts: opts} = hook,
+        attributes,
+        text_content,
+        meta
+      ) do
+    case opts[:bundle_mode] do
+      :runtime ->
+        hashed_name = ColocatedHook.hashed_name(meta)
+
+        new_content = """
+        window["phx_hook_#{hashed_name}"] = function() {
+          #{text_content}
+        }
+        """
+
+        {:keep, Map.put(attributes, "data-phx-runtime-hook", {:string, hashed_name}), new_content,
+         %{hashed_name: hashed_name}}
+
+      :current_otp_app ->
+        if ColocatedHook.should_bundle?(hook) do
+          {:drop, ColocatedHook.process_bundled_hook(hook, text_content, meta)}
+        else
+          {:drop, nil}
+        end
+
+      _ ->
+        {:drop, ColocatedHook.process_bundled_hook(hook, text_content, meta)}
+    end
+  end
+
+  def postprocess_tokens(
+        %Phoenix.LiveView.ColocatedHook{name: hook_name},
+        %{hashed_name: hashed_name},
+        tokens
+      ) do
+    Enum.map(tokens, fn
+      {:tag, name, attrs, meta} ->
+        {:tag, name, rewrite_hook_attrs(hook_name, hashed_name, attrs), meta}
+
+      {:local_component, name, attrs, meta} ->
+        {:local_component, name, rewrite_hook_attrs(hook_name, hashed_name, attrs), meta}
+
+      {:remote_component, name, attrs, meta} ->
+        {:remote_component, name, rewrite_hook_attrs(hook_name, hashed_name, attrs), meta}
+
+      other ->
+        other
+    end)
+  end
```

Comment:
One idea I had to not have to make the tokens public is to basically treat them as opaque and have a module with functions for working with them, e.g.

```elixir
  def postprocess_tokens(
        %Phoenix.LiveView.ColocatedHook{name: hook_name},
        %{hashed_name: hashed_name},
        tokens
      ) do
    alias Phoenix.LiveView.TagExtractorHelper

    TagExtractorHelper.map_tokens(tokens, fn token ->
      TagExtractorHelper.rewrite_attribute(token, "phx-hook", fn _old_value -> hashed_name end)
    end)
  end
```

Reply:
I went ahead with that approach.